File: programmers/고득점 Kit/DFSBFS/43163.py
# Solved (17m)
def check_one_diff(a, b):
    cnt = 0
    for x, y in zip(a, b):
        if x!=y:
            cnt += 1
            
    if cnt == 1:
        return True
    else:
        return False

def solution(begin, target, words):
    if target not in words:
        return 0
    
    result_cnt = 1e8
    visited = [False] * len(words)
    def dfs(cnt, word):
        nonlocal result_cnt
        if word == target:
            result_cnt = min(result_cnt, cnt)
        
        for idx, x in enumerate(words):
            if check_one_diff(word, x) and not visited[idx]:
                visited[idx] = True
                dfs(cnt+1, x)
                visited[idx] = False

    dfs(0, begin)
    return result_cnt

'''
하나만 다른지 찾는 방법? -> 함수 만듦
'''

# 다른 사람 풀이) Generator
from collections import deque

# 한 글자만 변경하여 만들 수 있는 단어들을 찾는 제너레이터 함수
def get_adjacent(current, words):
    for word in words:
        if len(current) != len(word):
            continue

        count = 0
        for c, w in zip(current, word):
            if c != w:
                count += 1

        if count == 1:
            yield word


def solution(begin, target, words):
    dist = {begin: 0}   # 각 단어까지의 변환 횟수를 저장
    queue = deque([begin])

    while queue:
        current = queue.popleft()

        for next_word in get_adjacent(current, words):
            if next_word not in dist:
                dist[next_word] = dist[current] + 1
                queue.append(next_word)

    return dist.get(target, 0)  # target 단어가 발견되면 변환 횟수를 반환하고, 불가능한 경우 0을 반환

'''
방문 여부를 사전형에 추가 되었는지 아닌지로 알 수 있다
'''

File: programmers/고득점 Kit/DFSBFS/43163.java
import java.util.*;
class Solution {
    int answer = (int)1e9;
    boolean[] visited = new boolean[50];
    public boolean check(String a, String b) {
        int cnt = 0;
        for(int i=0; i<a.length(); i++) {
            if(a.charAt(i)!=b.charAt(i)) cnt++;
        }

        if(cnt==1) return true;
        else return false;
    }
    public void dfs(int cur, int cnt, String target, String[] words) {
        if(words[cur].equals(target)) {
            answer = Math.min(answer, cnt);
            return;
        }

        visited[cur] = true;
        for(int i=0; i<words.length; i++) {
            if(!visited[i] && check(words[cur], words[i])) {
                dfs(i, cnt+1, target, words);
            }
        }
    }
    public int solution(String begin, String target, String[] words) {
        for(int i=0; i<words.length; i++) {
            if(check(begin, words[i])) {
                visited[i] = true;
                dfs(i, 1, target, words);
                visited[i] = false;
            }
        }
        return answer==1e9? 0 : answer;
    }
}

/*
if(words[cur]==target)
    ==는 객체의 주소 비교이므로 문자열 비교에 적절하지 않음
    내용 비교를 해야 하므로 equals()를 써야 함
 */

// 다른 사람 풀이
import java.util.*;
class Solution {
    
    class Word {
        String word;
        int depth;
        
        Word(String word, int depth) {
            this.word = word;
            this.depth = depth;
        }
    }
    
    public int solution(String begin, String target, String[] words) {
        boolean[] visited = new boolean[words.length];
        Queue<Word> queue = new LinkedList<>();
        
        queue.offer(new Word(begin, 0));
        
        while(!queue.isEmpty()) {
            Word current = queue.poll();
            
            if(current.word.equals(target)) {
                return current.depth;
            }
            
            for(int i = 0; i < words.length; i++) {
                if(!visited[i] && isOneDiff(current.word, words[i])) {
                    visited[i] = true;
                    queue.offer(new Word(words[i], current.depth + 1));
                }
            }
        }
        return 0;
    }
    
    private boolean isOneDiff(String a, String b) {
        int diff = 0;
        for(int i = 0; i < a.length(); i++) {
            if(a.charAt(i) != b.charAt(i)) diff++;
        }
        return diff == 1;
    }
}